enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+  <match **>
+     @type copy
+     #store section for each endpoint type
+     <store>
+       @type relabel
+       @label @SECUREFORWARD_OFFCLUSTER
+     </store>
+     <store>
+       @type relabel
+       @label @ELASTICSEARCH
+     </store>
+ </match>
+</label>
+ 
+#specific pipeline endpoint
+<label @ELASTICSEARCH>
+  <match retry_elasticsearch>
+     @type copy
+     <store>
+       @type elasticsearch
+       @id elasticsearch
+       client_key /var/run/ocp-collector/secrets/elasticsearch/key
+       client_cert /var/run/ocp-collector/secrets/elasticsearch/cert
+       ca_file /var/run/ocp-collector/secrets/elasticsearch/cacert
+       <buffer>
+		@type file
+		path '/var/lib/fluentd/elasticsearch'
+       </buffer>
+     </store>
+  </match>
+  <match **>
+     @type copy
+     <store>
+       @type elasticsearch
+       @id elasticsearch
+       client_key /var/run/ocp-collector/secrets/elasticsearch/key
+       client_cert /var/run/ocp-collector/secrets/elasticsearch/cert
+       ca_file /var/run/ocp-collector/secrets/elasticsearch/cacert
+       retry_tag "retry_elasticsearch"
+     </store>
+ </match>
+</label>
+<label @SECUREFORWARD_OFFCLUSTER>
+   ...
+</label>
+```
+
+Following is a diagram which identifies the workflow of a message through fluentd as captured by the above configuration.
+
+![https://docs.google.com/drawings/d/18FD4c1pBz4e8eJIH37yUwM40ew69g8h1cqPEfOLsuyY/edit?usp=sharing](images/log-forwarding-message-flow.png)
+
+### Cluster Logging Operator
+* The operator generates output configuration for the collector
+* The operator specifies in the collector daemonset the secrets needed to communicate to each endpoint:
+
+   `/var/run/ocp-collector/secrets/<secretName>/{cert,key,cacert}`
+
+    where:
+
+    * `secretName` is the name of a secret in the `openshift-logging` namespace.
+    * `cert`, `key`, `cacert` are files in the secret
+#### Collector
+* Collectors will be modified to be remove endpoint config specific logic from the start script ; configuration is assumed to be correct and used as provided by the cluster-logging-operator
+* Extract all configuration from after the source config point to final destination into the collector configuration.
+
+### Test Plan
+#### Unit testing
+* Log forwarding will add unit tests to provide adequate coverage for all changes
+* BDD unit testing will be added to unit testing to make tests goals more explicit,  readable, and obvious
+#### Integration and E2E tests
+* Test no regressions after upgrade when log forwarding exists
+* Tests to verify there are no regressions in functionality between pre and post logforwarding
+* Tests to verify log forwarding is writing logs to an Elasticsearch instance not managed by cluster logging
+* Tests to verify log forwarding is writing logs to an Elasticsearch instance managed by cluster logging
+* Tests to verify log forwarding is writing logs to a Fluentd instance via secure forward that is not managed by cluster logging
```

Comment:
We already have tests for most of these in oal - we might have to resurrect the es-copy test for the test of the Elasticsearch instance not managed by cluster logging

Reply:
Do we need to call out specific versions of Elasticsearch we support?  Or do we just want to say "we provide the latest Fluentd Elasticsearch plugin and the latest Ruby Elasticsearch libraries, which should work on all recent versions of Elasticsearch, but we cannot guarantee that they will work with every version of Elasticsearch"?

Reply:
> Do we need to call out specific versions of Elasticsearch we support? Or do we just want to say "we provide the latest Fluentd Elasticsearch plugin and the latest Ruby Elasticsearch libraries, which should work on all recent versions of Elasticsearch, but we cannot guarantee that they will work with every version of Elasticsearch"?

Not certain that needs to be in this document but we should document the versions against which we have tested and leave it at that.

Reply:
@richm @ewolinetz suggestions on replacing 'certficates' to something else?  `config`?  I believe keeping some root makes sense with `secretName` as the child.  These would allow us to enhance it in future as needed

Reply:
replace `certificates` with `secrets` - the value of `secrets` is the name of a secret.  The secret keys are the keys like `tls.cacert`, `tls.cert`, `tls.key`, `shared_key` - the very short list of keys

Reply:
in the global proxy config object it uses:
```
spec:
  . . .
  trustedCA:
    name: user-ca-bundle
```

Reply:
@ewolinetz so `trustedCA, cert, key`?  or `trustedCA, tls.cert, tls.key`

Reply:
@jcantrill that snippet is from the spec of the proxy config object. which was a response to https://github.com/openshift/enhancements/pull/16#discussion_r326731068.

For the contents of the secret, I think `tls.crt (note: lack of "e" to match the "tls" secret type names)` `tls.key` for sure. and maybe something like `additionalTrustBundle".

@bparees what is the field name used in `secret/user-ca-bundle`?

Reply:
not sure what `secret/user-ca-bundle` is?

closest thin
g i can think of that you are referring to is the `configmap/trusted-ca-bundle` which uses `ca-bundle.crt` as the key.

or `configmap/service-ca` which also uses `ca-bundle.crt`.

and then there's the configmaps you create and annotate with `service.beta.openshift.io/inject-cabundle: true` which get a key named `service-ca.crt` injected....

Reply:
oh and there's the secrets you ask for/get for server signing certs..... that secret will contain a `tls.crt` and `tls.key` key.

Reply:
@jcantrill i'd suggest `ca-bundle.crt`, `tls.crt`, and `tls.key` then as field names